packages/components/docs/immer.md

```diff
@@ -0,0 +1,235 @@
+
+1. Immutability with normal JavaScript objects, arrays, Sets and Maps. No new APIs to learn!
+1. Strongly typed, no string based paths selectors etc.
+1. Structural sharing out of the box
+1. Object freezing out of the box
+1. Deep updates are a breeze
+1. Boilerplate reduction. Less noise, more concise code.
+1. First class support for patches
+1. Small: 3KB gzipped
+
+## Learning Immer
+
+This document intends to provide some specific insights about using Immer in our LabKey client-side code. As such, it expects
+the reader to have an understanding of why Immer exists, what Immer does, and how Immer does it.
+
+Before you read further it is **highly recommended** that you read (or watch) at least one the following:
+
+- [Immer's official documentation](https://immerjs.github.io/immer/docs/introduction) - read the docs! Most up-to-date and covers lots of topics.
+- [Introducing Immer: Immutability the easy way](https://hackernoon.com/introducing-immer-immutability-the-easy-way-9d73d8f71cb3) - written by the author of Immer
+- [Simplify Creating Immutable Data Trees With Immer](https://egghead.io/lessons/redux-simplify-creating-immutable-data-trees-with-immer) - egghead.io video tutorial
+
+## Scenarios
+
+This section focuses on a couple of scenarios to help get a better understanding. These were written against `v6.0.3` of
+Immer so things may have changed if you're working with a more current version.
+
+### Immutable class
+
+This scenario highlights declaring an immutable class in TypeScript using Immer. By the end we'll have an immutable class
+that is both compile-time and run-time safe. To keep the class simple we're going to define a `Circle` class defined only
+by it's `radius`:
+
+```ts
+class Circle {
+    radius: number;
+
+    constructor(r: number) {
+        this.radius = r;
+    }
+}
+```
+
+#### Run-time safety
+
+This initial declaration is fully mutable. You can externally modify the radius after construction.
+
+```ts
+let circle = new Circle(5);
+circle.radius = 10; // radius now 10
+```
+
+Let's try to use Immer on this class and see if it works:
+
+```ts
+import { produce } from 'immer';
+
+let circle = produce(new Circle(5), () => {});
+circle.radius = 10; // circle.radius is now 10! I thought using produce made it immutable!
+```
+
+To make a class immutable with Immer you first annotate the class with a [Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) called
+provided by Immer called `immerable`:
+
+```ts
+import { immerable } from 'immer';
+
+class Circle {
+    [immerable] = true;
+
+    radius: number;
+
+    constructor(r: number) {
+        this.radius = r;
+    }
+}
+```
+
+What does this symbol do? To quote the Immer docs:
+
+> Every other object must use the `immerable` symbol to mark itself as compatible with Immer.
+When one of these objects is mutated within a producer, its prototype is preserved between copies.
```

Comment:
This seems like it could use a little more context what "Every other object" means.  Maybe including the previous line like

Plain objects, arrays, Maps and Sets are always [compatible with] Immer.

Reply:
Thanks for the suggestion. I switched it to a paraphrase and replaced "Every other object" with "Classes".